### ISSUE TITLE
docker client doesn't pass signals when a terminal is attached

### ISSUE THREAD

**Issue by jarro2783** [moby/moby#28872]
**Description**

Re #3793: when running with `-ti`, `docker run` does not pass signals to child process.

**Steps to reproduce the issue:**
1. I built an image `FROM base/archlinux`, with https://github.com/krallin/tini in it.
2. `docker run -ti tini-arch /tini -- sleep 300`
3. `ps -A f | grep docker`

        31807 pts/5    Sl+    0:00              |   |   |   \_ docker run -ti tini-arch /tini -- sleep 300

4. `kill 31807`

**Describe the results you received:**

In the terminal window attached I see

    Terminated

In another window:
```
docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS              PORTS               NAMES
94e81d8dd7db        tini-arch           "/tini -- sleep 300"   59 seconds ago      Up 58 seconds                           angry_ardinghelli
```

**Describe the results you expected:**

I expect the container to stop running because it has been passed the appropriate signal. I have confirmed that running the same without `-ti` prints nothing in the attached terminal, and `docker ps` shows that the running container has stopped.

This is necessary because I expect that in situations like running a container over an SSH session should terminate the container if the SSH session hangs up, because SIGHUP will be sent to the client when the terminal closes.

**Output of `docker version`:**

```
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.6.1
 Git commit:   4dc5990
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.6.1
 Git commit:   4dc5990
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 19
 Running: 0
 Paused: 0
 Stopped: 19
Images: 206
Server Version: 1.11.0
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.6-gentoo
Operating System: Gentoo/Linux
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.829 GiB
Name: guepardo
ID: Y6HO:RPCY:TMYW:63IN:CQP2:I4JY:6EUL:AYHK:HBH2:JK7U:YCYJ:ODTY
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Username: jarro2783
Registry: https://index.docker.io/v1/
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical machine. I have confirmed this on another physical host running 1.12.0.

**Comment by jarro2783** [comment]
This also affects `docker exec`. The effect is worse because there is no way to reattach (or even kill if you're container doesn't support it) the process that is left running.

**Comment by thaJeztah** [comment]
I think this works as expected; the `docker` command line tool, and the docker _daemon_ (`dockerd` in 1.12 and up) are separate processes; the client talks with the daemonn through the remove API.

The problem you're having is that you are killing the _client_ process that started the container. Nothing happens with the container (or it's processes) because of that, only the client disconnects.

You can see that if you look at the _full_ output of your `ps` command;

```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
...
root      1604  0.0  0.3  95368  6980 ?        Ss   10:07   0:00  \_ sshd: root@pts/0
root      1690  0.0  0.2  21300  5188 pts/0    Ss   10:07   0:00      \_ -bash
root     16840  0.2  1.3 251916 28044 pts/0    Sl+  10:14   0:00          \_ docker run -it alpine sleep 3000
root      1636  0.0  0.2  36800  4484 ?        Ss   10:07   0:00 /lib/systemd/systemd --user
root      1639  0.0  0.1  61492  2184 ?        S    10:07   0:00  \_ (sd-pam)
root     16489  0.7  2.3 507468 49080 ?        Ssl  10:12   0:01 /usr/bin/docker daemon -H fd://
root     16497  0.0  0.6 287256 13608 ?        Ssl  10:12   0:00  \_ docker-containerd -l /var/run/docker/libcontainerd/docker-containerd.sock --runtime docker-runc --start-timeout 2m
root     16863  0.0  0.2 200752  5416 ?        Sl   10:14   0:00      \_ docker-containerd-shim 99680cddeec49a4347bc1bad0a1c8e694e6b8b62766c7e3e12b611b4dc8bca50 /var/run/docker/libcontainerd/99680cddeec49a4347bc1bad0a1c8e694e6b8b62766c7e3e12b611b4dc8bca50 docker-runc
root     16905  0.2  0.0   1508     4 pts/1    Ss+  10:14   0:00          \_ sleep 3000
```

You can see that the `sleep` process is _not_ a child process of `docker run -it alpine sleep 3000`. The `docker run -it alpine sleep 3000` could even be on a separate host if you're connecting with a remote daemon.

And after killing the client;

```
root      1601  0.0  0.3  65520  6168 ?        Ss   10:07   0:00 /usr/sbin/sshd -D
root     16743  0.0  0.3  95368  6924 ?        Ss   10:13   0:00  \_ sshd: root@pts/2
root     16776  0.0  0.2  21296  5192 pts/2    Ss   10:13   0:00      \_ -bash
root     17002  0.0  0.1  36196  3452 pts/2    R+   10:21   0:00          \_ ps auxf
root     16489  0.2  2.2 507468 45852 ?        Ssl  10:12   0:01 /usr/bin/docker daemon -H fd://
root     16497  0.0  0.6 287256 12912 ?        Ssl  10:12   0:00  \_ docker-containerd -l /var/run/docker/libcontainerd/docker-containerd.sock --runtime docker-runc --start-timeout 2m
root     16863  0.0  0.2 200752  5436 ?        Sl   10:14   0:00      \_ docker-containerd-shim 99680cddeec49a4347bc1bad0a1c8e694e6b8b62766c7e3e12b611b4dc8bca50 /var/run/docker/libcontainerd/99680cddeec49a4347bc1bad0a1c8e694e6b8b62766c7e3e12b611b4dc8bca50 docker-runc
root     16905  0.0  0.0   1508     4 pts/1    Ss+  10:14   0:00          \_ sleep 3000
```

> The effect is worse because there is no way to reattach (or even kill if you're container doesn't support it) the process that is left running.

The process still shows up as usual after killing the client (see my above example), and the container is still there, so you can  `docker stop <your container>` or `docker kill <your container>`;

```
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
99680cddeec4        alpine              "sleep 3000"        9 minutes ago       Up 9 minutes                            high_snyder
```

**Comment by jarro2783** [comment]
No, this is not expected at all. The fact that the client and the daemon are separate is exactly the point of my issue. If I am connected to a host via ssh, and my terminal hangs up due to a lost connection, then as per Linux, the client process will receive SIGHUP. My expectation is that docker preserves this behaviour of Linux, and has the client tell the daemon that it got a SIGHUP, which then tells the child process. Otherwise there is no way that a terminal hangup will ever be propagated into the container.

In the case of `exec`, what I want to do is keep the container running, but run another process inside it using `docker exec`. Then if my terminal hangs up I want just the `exec`ed process to die, I do not want to then have to kill the whole container.

**Comment by thaJeztah** [comment]
Ah, alright, I misinterpreted your description, and thought you assumed you killed the _container_ process.

I digged into the history for this, and it looks to be [purposefully disabled if a TTY is enabled](https://github.com/docker/docker/blob/eb131c5383db8cac633919f82abad86c99bffbe5/cli/command/container/run.go#L129-L132). See https://github.com/docker/docker/pull/2426#issuecomment-27278150. Removing that part of the code seems to result in what you are describing.

It's a _long_ time ago though that that code was added; let me ping @tianon because he may remember the exact reason there.

**Comment by jarro2783** [comment]
That's a bit curious then. [Two comments later](https://github.com/docker/docker/pull/2426#issuecomment-27324597) seems to indicate that it doesn't work very well though.

I just thought I would add too that this seems to work fine when a TTY is not enabled, so none of the issues in #3793 are the problem here.

These are the processes *inside* the container, so `tini` will pass signals as expected:
```
14707 pts/18   Ss     0:00          \_ /tini sleep 400
14719 pts/18   S+     0:00              \_ sleep 400
```

Then the client is this:
```
15403 pts/16   Sl+    0:00              \_ docker run -ti tini-arch /tini sleep 400
```
which when killed leaves the container running, if it was started with a terminal.

**Comment by tianon** [comment]
I don't remember the reasoning behind this. :disappointed: 

Does it work as expected if an explicit `--sig-proxy` is specified as well?

**Comment by thaJeztah** [comment]
Looks like that code overrides it at all times, but I can double check

**Comment by jarro2783** [comment]
No it doesn't work with `--sig-proxy`. I modified my local docker, commenting out those lines, and it still doesn't work. So maybe there is something else going on as well.

On a related note, if the docker client were to crash, or be `kill -9`ed, I feel like having a terminal connected should also result in a SIGHUP, since that is effectively what has happened.

Also can someone explain to me what `-i` is supposed to actually do? The help for `-t` seems to indicate that it should work by itself, but I definitely don't get an interactive terminal without `-i`. However the help for `-i` seems to indicate that if I don't use it, I will get the behaviour that I want, because I do not want to `keep STDIN open even if not attached`.

**Comment by hjdr4** [comment]
Same problem here when running a shell as entrypoint through ssh, using Docker 1.13.
I have a container running bash as entry point.
The 'docker run' command is wrapped into a shell script doing 'exec docker run -it myCustomImageWithBashAsEntryPoint'

If I setup a remote ssh server to run my 'exec docker run -it' wrapper as shell, I fall into the dockerized shell when connecting through ssh.
If I close the ssh client, I expect the container to be stopped.
Actually it does not stop the container. it just kills the 'docker run' command.

As said by someone before, I'm also confused with -i and -t options. I don't want to keep STDIN open if not attached. But I can't have a shell working without -i.

**Comment by hjdr4** [comment]
Those lines in **cli/command/container/run.go** look to be the issue :

```//Disable sigProxy when in TTY mode
if config.Tty {
      opts.sigProxy = false
}
```

I have build a client without those lines and I get the expected behavior.

**Comment by crosbymichael** [comment]
Doesn't the tty handle forwarding signals to the container process in this case?

**Comment by jarro2783** [comment]
No. The `daemon` has a different `tty` to the client process. So `tty` sends the signal to the client process, and it just dies. The daemon keeps the container process running because it doesn't disconnect the container's tty when the client dies.

**Comment by hashar** [comment]
This is an exact duplicate of #9098 . The reason is using `--tty` causes signal proxying to be disabled (even if one also uses `--sig-proxy`), `docker run` and `docker exec` sharing the same code path that has that behavior.

I would recommend to close this #28872 as a dupicate of #9098.

**Comment by jarro2783** [comment]
This does appear to be the same thing. Having `--tty` disable signal proxying seems very bizarre to me. If you ask for a `tty`, you are expecting consistency with Linux terminal behaviour: a SIGHUP when the controlling terminal hangs up, which is exactly what has happened in this case.

**Comment by Mark-50** [comment]
@hashar this is not the same situation with #9098.
when "docker run" process was killed, the container started by "docker run" should exit.
when "docker exec" process was killed, the container started before should remain running, only the process started by "docker exec" inside the container should exit.

also, this is not only about signal proxying.
what if we run "docker exec" or "docker run -H " in a host and dockerd was started in a remote host?
signal proxy is not sufficient for this situation.

**Comment by hashar** [comment]
@Mark-50 wrote:
> @hashar this is not the same situation with #9098.
> when "docker run" process was killed, the container started by "docker run" should exit.
> when "docker exec" process was killed, the container started before should remain running, only the process started by "docker exec" inside the container should exit.

I think I understood the code for docker run and docker exec to both hit the same faulty code path. And I am pretty sure at time (back in 2017) I did reproduce the same behavior for both run and exec subcommands.

Anyway for the docker run case, I am pasting here my comment from the other issue https://github.com/moby/moby/issues/9098#issuecomment-347536699 :


I had that issue with both `docker run` and `docker exec` not dispatching signals to the daemon/container. The root cause is when using `--tty` signal proxying is entirely disabled with no way to enable it (even with `--sig-proxy`).  It affects at least `docker run` and `docker exec` that share the same code path.

Previously `--sig-proxy` was an option to force proxying signals when not using a tty, and hence when using `--tty` the proxy got forwarded. Passing both `--sig-proxy` and `-tty` was leading to an error. That looked like:

Options | Signal proxy?
--- | ---
_none_ | No
`--tty` | Yes
`--sig-proxy` | Yes
`--tty --sig-proxy` | error: TTY mode (-t) already imply signal proxying (-sig-proxy)

October 2013 patch e0b59ab52b87b8fc15dd5534c3231fdd74843f9f was made to "_Enable sig-proxy by default in run and attach _". It changed `--sig-proxy` to default to true and made `--tty` to always disable signal proxy.

Options | Signal proxy?
--- | ---
_none_ | Yes (was No)
`--tty` | No (was Yes)
`--sig-proxy` | Yes
`--tty --sig-proxy` | No (was an error)

So what happened with e0b59ab52b87b8fc15dd5534c3231fdd74843f9f is that signal proxying is now enabled by default for non-tty. BUT the patch has a fault: setting `--tty` always disable signal proxying.

I am pretty sure that signal-proxying should be enabled by default whether it is a tty or non-tty mode, it is still possible to disable it with `--sig-proxy=false`. A patch would thus have to implement the following changes:


Options | Signal proxy (current) | Expected
--- | --- | ---
_none_ | Yes | Yes
`--tty` | No | **Yes**
`--sig-proxy` | Yes | Yes
`--sig-proxy=false` | No | No
`--tty --sig-proxy` | No | **Yes**
`--tty --sig-proxy=false` | No | No

TLDR: `--tty` should not arbitrarily force `sigProxy = false` caused by e0b59ab52b87b8fc15dd5534c3231fdd74843f9f

Reference: https://phabricator.wikimedia.org/T176747#3749436

**Comment by Dragas** [comment]
Seems not to be limited to using `tty`. Docker doesn't pass `^C` and `^D` signals while running commands in `docker run` at all.

**Comment by hashar** [comment]
One would have to look at the code, but if I remember correctly, `docker run` and `docker exec` ends up hitting the same code and are thus equally broken.


The fix is really straightforward: `--tty` should not stop forwarding signal forwarding as per my previous comment that list the truth table.  Unfortunately I do not know much about golang and really nothing about the Docker code base or how to write unit tests in go :-\

**Comment by cpuguy83** [comment]
So in the docker cli code there is an explicit check for `--tty` and in such cases it disables signal proxying.
I do not think anyone knows why this exists, but it's existed for a very long time.

See docker/cli#1841

**Comment by thaJeztah** [comment]
The fix itself was easy, but I'm failing to get the test to work; https://github.com/docker/cli/pull/1841

didn't have time to look at it recently, but help on getting the test working is appreciated.

**Comment by thaJeztah** [comment]
This should be fixed by https://github.com/docker/cli/pull/1841 / https://github.com/docker/cli/pull/2177, which is in docker 19.03.5 and up

**Comment by hashar** [comment]
@thaJeztah thank you very much for the sig-proxy. Being able to Ctrl+C has really improved my user experience ;)

**Comment by thaJeztah** [comment]
Good to hear it fixed your issue 👍 🥳
